### ISSUE TITLE
Wrong SQL when filtering by ManyHasMany property

### ISSUE THREAD

**Issue by step0s** [nextras/orm#390]
**Describe the bug**

When filtering in Collection by two ManyHasMany properties that points to the same table, generated SQL is wrong.

**To Reproduce**

I have a table `log` that has relations to tables `group` and `message`. `message` has relation to `group` and `group` has many members in table `person`. When I filter all logs that have relation to groups whose member is a certain person like this, everything works fine:
```
$logs = $this->orm->logRepository->findBy([
    'this->group->persons->id' => $this->loggedPerson->id,
]);
```
However when I also want to show logs with all messages from groups whose member is certain person, the generated SQL is wrong:
```
$logs = $this->orm->logRepository->findBy([
    ICollection::OR,
    'this->group->persons->id' => $this->loggedPerson->id,
    'this->message->group->persons->id' => $this->loggedPerson->id,
]);
```
**Expected behavior**

This generates SQL
```
SELECT `log`.*
FROM `log` AS `log`
LEFT JOIN `group` AS `group` ON (`log`.`group_id` = `group`.`id`)
LEFT JOIN `group2person` AS `group2person` ON (`message_group`.`id` = `group2person`.`group_id`)
LEFT JOIN `person` AS `group_persons` ON (`group2person`.`person_id` = `group_persons`.`id`)
LEFT JOIN `message` AS `message` ON (`log`.`message_id` = `message`.`id`)
LEFT JOIN `group` AS `message_group` ON (`message`.`group_id` = `message_group`.`id`)
LEFT JOIN `person` AS `message_group_persons` ON (`group2person`.`person_id` = `message_group_persons`.`id`)
WHERE ((`group_persons`.`id` = 1)) OR ((`message_group_persons`.`id` = 1))
GROUP BY `log`.`id`
```
while expected SQL is
```
SELECT `log`.*
FROM `log` AS `log`
LEFT JOIN `group` AS `group` ON (`log`.`group_id` = `group`.`id`)
LEFT JOIN `group2person` AS `group2person` ON (`group`.`id` = `group2person`.`group_id`)
LEFT JOIN `person` AS `group_persons` ON (`group2person`.`person_id` = `group_persons`.`id`)
LEFT JOIN `message` AS `message` ON (`log`.`message_id` = `message`.`id`)
LEFT JOIN `group` AS `message_group` ON (`message`.`group_id` = `message_group`.`id`)
LEFT JOIN `person` AS `message_group_persons` ON (`group2person`.`person_id` = `message_group_persons`.`id`)
WHERE ((`group_persons`.`id` = 1)) OR ((`message_group_persons`.`id` = 1))
GROUP BY `log`.`id`
```
Difference is in ON clause of second JOIN.

**Versions**

 - Database: MariaDB 10.3
 - Orm: Orm 3.1
 - Dbal: Dbal 3.1

**Comment by hrach** [comment]
hi, could you please post definition of your tables? at least all foreign keys. thanks

**Comment by step0s** [comment]
log
```
`id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
`group_id` INT(10) UNSIGNED NULL DEFAULT NULL,
`message_id` INT(10) UNSIGNED NULL DEFAULT NULL,
PRIMARY KEY (`id`),
CONSTRAINT `fk_log_group` FOREIGN KEY (`group_id`) REFERENCES `group` (`id`) ON UPDATE CASCADE ON DELETE CASCADE,
CONSTRAINT `fk_log_message` FOREIGN KEY (`message_id`) REFERENCES `message` (`id`) ON UPDATE CASCADE ON DELETE CASCADE
```
message
```
`id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
`group_id` INT(10) UNSIGNED NULL DEFAULT NULL,
PRIMARY KEY (`id`),
CONSTRAINT `fk_message_group` FOREIGN KEY (`group_id`) REFERENCES `group` (`id`) ON UPDATE CASCADE ON DELETE CASCADE
```
group
```
`id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
PRIMARY KEY (`id`)
```
group2person
```
`person_id` INT(10) UNSIGNED NOT NULL,
`group_id` INT(10) UNSIGNED NOT NULL,
PRIMARY KEY (`person_id`, `group_id`),
CONSTRAINT `fk_group2person_group` FOREIGN KEY (`group_id`) REFERENCES `group` (`id`) ON UPDATE CASCADE ON DELETE CASCADE,
CONSTRAINT `fk_group2person_person` FOREIGN KEY (`person_id`) REFERENCES `person` (`id`) ON UPDATE CASCADE ON DELETE CASCADE
```
person
```
`id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
PRIMARY KEY (`id`)
```

**Comment by hrach** [comment]
If I'm not mistaken, the expected query should be:

```SQL
SELECT `log`.*
FROM `log` AS `log`
LEFT JOIN `group` AS `group` ON (`log`.`group_id` = `group`.`id`)
LEFT JOIN `group2person` AS `group2person` ON (`group`.`id` = `group2person`.`group_id`)
LEFT JOIN `person` AS `group_persons` ON (`group2person`.`person_id` = `group_persons`.`id`)
LEFT JOIN `message` AS `message` ON (`log`.`message_id` = `message`.`id`)
LEFT JOIN `group` AS `message_group` ON (`message`.`group_id` = `message_group`.`id`)
LEFT JOIN `group2person` AS `message_group2person` ON (`message_group`.`id` = `message_group2person`.`group_id`)
LEFT JOIN `person` AS `message_group_persons` ON (`message_group2person`.`person_id` = `message_group_persons`.`id`)
WHERE ((`group_persons`.`id` = 1)) OR ((`message_group_persons`.`id` = 1))
GROUP BY `log`.`id`
```

E.g. one another group2person join is also missing. This is obviously an error in alias generation and as a result the QueryBuilder drops the another LEFT JOIN.

**Comment by hrach** [comment]
I've fixed this in branches:
- from master - https://github.com/nextras/orm/tree/alias_fix
- from v3.1 - https://github.com/nextras/orm/tree/alias_fix_v31

Please update orm via composer to `"nextras/orm": "dev-alias_fix_v31"` and comment here if this is fixed, thanks :)

Also thanks for the report of such shameful bug. 🙈

**Comment by step0s** [comment]
Thank you for the quick fix! It's working fine now :-)

**Comment by hrach** [comment]
Currently the fix is available in `dev-v3.1` and `master`, will release soon.
